x/stakeibc/keeper/reward_converter.go

```diff
@@ -0,0 +1,442 @@
+		Token:            sendTokens,
+		Sender:           tradeIcaAddress,
+		Receiver:         withdrawlIcaAddress,
+		TimeoutTimestamp: timeout,
+		Memo:             "",
+	})
+
+	hostZoneId := route.TradeToHostHop.ToAccount.ChainId
+	tradeZoneId := route.TradeToHostHop.FromAccount.ChainId
+	k.Logger(ctx).Info(utils.LogWithHostZone(hostZoneId,
+		"Preparing MsgTransfer of %+v from %s to %s", sendTokens, tradeZoneId, hostZoneId))
+
+	// Send the ICA tx to kick off transfer from hostZone through rewardZone to the tradeZone (no callbacks)
+	tradeZoneConnectionId := route.TradeToHostHop.FromAccount.ConnectionId
+	err := k.SubmitICATxWithoutCallback(ctx, tradeZoneConnectionId, types.ICAAccountType_TRADE, msgs, timeout)
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrICATxFailed, "Failed to submit ICA tx, Messages: %+v, err: %s", msgs, err.Error())
+	}
+
+	return nil
+}
+
+// Trade reward tokens in the Trade ICA for the target output token type using ICA remote tx on trade zone
+// The amount represents the total amount of the reward token in the trade ICA found by the calling ICQ
+// Depending on min and max swap amounts set in the route, it is possible not the full amount given will swap
+func (k Keeper) TradeRewardTokens(ctx sdk.Context, amount sdk.Int, route types.TradeRoute) error {
+	// If the min swap amount was not set it would be ZeroInt, if positive we need to compare to the amount given
+	//  then if the min swap amount is greater than the current amount, do nothing this epoch to avoid small swaps
+	if route.MinSwapAmount.IsPositive() && route.MinSwapAmount.GT(amount) {
+		return nil
+	}
+
+	// If the max swap amount was not set it would be ZeroInt, if positive we need to compare to the amount given
+	//  then if max swap amount is LTE to amount full swap is possible so amount is fine, otherwise set amount to max
+	if route.MaxSwapAmount.IsPositive() && route.MaxSwapAmount.GT(amount) {
+		amount = route.MaxSwapAmount
+	}
+
+	// See if pool swap spot price has been set to a valid ratio (string representing a float like "10.203")
+	// If there is a valid spot price, use it to set a floor for the acceptable minimum output tokens
+	// 5% slippage is allowed so multiply by 0.95 the expected spot price to get the target minimum
+	// minOut is the minimum number of route.TargetDenomOnTradeZone we must receive or the swap will fail
+	minOut := sdk.ZeroInt()
+	if route.SpotPrice != "" {
+		if spotPrice, err := sdk.NewDecFromStr(route.SpotPrice); err == nil {
+			slippageRatio := sdk.NewDecWithPrec(95, 2) // 0.95 to allow 5% loss to slippage
+			inputToOutputRatio := slippageRatio.Mul(spotPrice)
+			minOut = sdk.NewDecFromInt(amount).Mul(inputToOutputRatio).TruncateInt()
+		} else {
+			k.Logger(ctx).Error("Couldn't parse the spot price %s as a Decimal", route.SpotPrice)
+			// Don't allow a missing spot price to stop the swap, use zeroInt as the lower bound for now
```

Comment:
we'd only not have a price immediately after the TradeRoute is registered right? Feels like we could go ahead and throw an error here since this is unexpected

Reply:
Yes true, although as we discussed, it could also happen that they change something on their side so queries break somehow and the data we have becomes out of date, or in the wrong format.  We want to add a lastUpdatedTimestamp so we know how fresh the price is.

Reply:
yeah sorry my comment was a bit stale. As we discussed on the call/slack, we'll just keep executing the trade even if the price is stale. This is implemented correctly in this branch, marking resolved